### DISCUSSION TITLE
`@ObservedObject` への DI を `onAppear` で決して行わない。

### DISCUSSION THREAD

**Discussion by YusukeHosonuma** [YusukeHosonuma/Effective-SwiftUI#26] [Pending]
## Overview

SwiftUI において、`ObservableObject` に任意のオブジェクトを DI したいケースが存在する。

例えば、以下のコードでは `@EnvironmentObject` 経由で取得した `AuthState` を `RootViewModel` に引き渡そうとしている。

```swift
class AuthState: ObservableObject {
   init() {}
}

struct ContentView: View {
    var body: some View {
        RootView()
            .environmentObject(AuthState())
    }
}

struct RootView: View {
    @EnvironmentObject var authState: AuthState // 💡 Obtained via `@EnvironmentObject`.
    @ObservedObject var viewModel: RootViewModel
    
    init() {
        // 🚫 Compile error: 'self' used before all stored properties are initialized.
        viewModel = RootViewModel(authState: authState)
    }
    
    var body: some View { ... }
}
    
class RootViewModel: ObservableObject {
    private let authState: AuthState // 💡 RootViewModel required `AuthState`.
    
    init(authState: AuthState) {
        self.authState = authState
    }
}
```

しかし、コメントで記載しているとおり、上記のコードはコンパイルエラーとなる。

Initializer において `@EnvironmentObject` で宣言された `authState` がプロパティが初期化されていない段階でアクセスしようとしているためだ。これは SwiftUI ではなく、Swift 言語における言語仕様からくる制限だ。

では、以下のようなコードはどうだろうか？

```swift
struct RootView: View {
    @EnvironmentObject var authState: AuthState

    // 🚫 Compile error:
    // Cannot use instance member 'authState' within property initializer; 
    // property initializers run before 'self' is available
    @ObservedObject var viewModel = RootViewModel(authState: authState)
    
    var body: some View { ... }
}
```

これも本質的にはさきほどのコードと変わらず、初期化が完了していない時点で `authState` プロパティにアクセスしようとしているためコンパイルエラーとなる。

では、`lazy` を使用して初期化を遅延させたらどうだろう？

```swift
struct RootView: View {
    @EnvironmentObject var authState: AuthState

    // 🚫 Compile error: Property 'viewModel' with a wrapper cannot also be lazy
    @ObservedObject lazy var viewModel = RootViewModel(authState: authState) // 💡 Try `lazy` keyword.
    
    var body: some View {}
}
```

これも残念ながらコンパイルエラーとなる。Property-wrapper に対して `lazy` キーワードは使用できないためだ（そもそも Property-wrapper の時点で、ある種の `lazy` を実現しているとも言える）。

それでは考え方を変えて、画面が表示された際に呼ばれる `onAppear` のタイミングにおいて、メソッド経由で DI （method injection）する方法はどうだろうか。

```swift
struct RootView: View {
    @EnvironmentObject var authState: AuthState
    @ObservedObject var viewModel = RootViewModel()
    
    var body: some View {
        Text("Hello")
            .onAppear {
                viewModel.inject(authState) // 🚫 Should definitely avoided!!
            }
    }
}

class RootViewModel: ObservableObject {
    private var authState: AuthState!
    
    // 💡 DI by method injection.
    func inject(_ authState: AuthState) {
        self.authState = authState
    }
}
```

これはコンパイルこそパスするものの、絶対に避けるべき実装パターンである。

一見するとこれは正しく動作するし、実際に上記のコードを動かしてみても問題は見つけられない。

しかし、View の初期化タイミングと `onAppear()` のタイミングは常に一致するわけではないため、実行時に以下のようなことが起こりうる。

1. `RootView` の再生成によって `RootViewModel` が再初期化される。
2. `onAppear` が呼ばれず、`RootViewModel` の `AuthState` にオブジェクトが設定されない。
3. `RootViewModel` は不正な状態のまま機能し続けようとする。

上記のコード例では `authState` を強制アンラップ型（`AuthState!`）で宣言しているため、内部的に `authState` にアクセスしようとしたタイミングでクラッシュを引き起こす。

Optional型（`AuthState?`）にすればクラッシュこそ避けられるものの、不正なステートのまま動作しているため、結局のところ期待しない動作をすることに代わりはない。

こうした意図しない不正なステートを発生させないため、`@ObservedObject` への DI を `onAppear` などで決して行ってはならない。

## Discussion

この問題が発生する具体的なコードを忘れてしまったため、正確な記事を書けなくなってしまった；

以前、この問題を解決する最もシンプルな手段が`@ObservedObject` の代わりに `@StateObject` を利用することだと考えたのだけれど、結局のところ不正なステートを一時的にしろ発生させるので、`onAppear` などで DI を行うこと自体避けるべきパターンなのかもしれない。

```swift
struct RootView: View {
    @EnvironmentObject var authState: AuthState
    @StateObject var viewModel = RootViewModel() // ✅ Use `@StateObject`
    
    var body: some View {
        Text("Hello")
            .onAppear {
                viewModel.inject(authState)
                viewModel.onAppear()
            }
    }
}

class RootViewModel: ObservableObject {
    private var authState: AuthState!
    
    func inject(_ authState: AuthState) {
        self.authState = authState
    }
    
    func onAppear() {
        // 💡 Use `authState`
    }
}
```

あるいは `onAppear` で初期化されるまで、絶対に `authState` にアクセスしないように実装する手もあるかもしれないが、簡単にミスを引き起こしそうな気もしなくもない（`Optional` にすればクラッシュは確実に防げるが...）。

**Comment by uhooi** [comment]
話が変わるかもしれませんが、 `@StateObject` も View のイニシャライザで初期化できるようになりました↓

```swift
struct SakatsuInputScreen: View {
  @StateObject private var viewModel: SakatsuInputViewModel<SakatsuUserDefaultsClient>

  init(sakatsu: Sakatsu?) {
    self._viewModel = StateObject(wrappedValue: SakatsuInputViewModel(sakatsu: sakatsu ?? .init()))
  }
}
```

ref: https://github.com/uhooi/Loki/blob/5a56c84f1318cfe03c10f2b7b911a6e3fbd7560e/LokiPackage/Sources/Features/Sakatsu/SakatsuInput/SakatsuInputScreen.swift

ViewModel に型パラメータがある場合、View 側で具象型を指定する必要があるのが気になりますが、、
